### PR TITLE
Enable es6toprimitive under experimental

### PR DESCRIPTION
**Pull request by suwc**
Enable es6toprimitive under experimental flag.
Update following areas to be spec-compliant:

ToPrimitive (abstract operation)
Date.prototype[@@toprimitive]()
Symbol.prototype[@@toprimitive]()

Update unit tests with coverage on:

ToNumber (abstract operation)
ToString (abstract operation)
ToPropertyKey (abstract operation)
Abstract rational comparison (abstract operation)
Abstract equality comparison (abstract operation)
"+" operator
Date() constructor
Date.prototype.toJSON()

### CODE REVIEW

**Comment by suwc** [comment]
@Microsoft/chakra-es could you take look pls?

**Comment by akroshg** [comment]
>             Assert(!CrossSite::NeedMarshalVar(result, requestContext));

nit: (not related to your change) this assert should be moved below if check and its body. If this result is nullptr then this assert will throw, isn't it?

---
Refers to: lib/Runtime/Language/JavascriptConversion.cpp:528 in bf80652. [](commit_id = bf8065293e0ef016de1c59275bac3ed82135ce34, deletion_comment = False)

File: test/es6/toPrimitive.js [hunk 1]
```diff
@@ -44,8 +44,33 @@ var tests = [
...
+            assert.areEqual(Symbol.hasInstance, Symbol.hasInstance[Symbol.toPrimitive](), "Symbol.hasInstance");
+            assert.areEqual(Symbol.isConcatSpreadable, Symbol.isConcatSpreadable[Symbol.toPrimitive](), "Symbol.isConcatSpreadable");
+            assert.areEqual(Symbol.match, Symbol.match[Symbol.toPrimitive](), "Symbol.match");
+            assert.areEqual(Symbol.replace, Symbol.replace[Symbol.toPrimitive](), "Symbol.replace");
+            assert.areEqual(Symbol.search, Symbol.search[Symbol.toPrimitive](), "Symbol.search");
+            assert.areEqual(Symbol.split, Symbol.split[Symbol.toPrimitive](), "Symbol.split");
+            assert.areEqual(Symbol.toStringTag, Symbol.toStringTag[Symbol.toPrimitive](), "Symbol.toStringTag");
+            assert.areEqual(Symbol.unscopables, Symbol.unscopables[Symbol.toPrimitive](), "Symbol.unscopables");
+
+            assert.throws(()=>Symbol.prototype[Symbol.toPrimitive].call(undefined), TypeError, "undefined", "Symbol[Symbol.toPrimitive]: 'this' is not a Symbol object");
```
Comment (akroshg) [inline test/es6/toPrimitive.js]: these are duplicated? same tests are there in the es6symbol.js, aren't they?

File: lib/Runtime/Library/JavascriptDate.cpp [hunk 2]
```diff
@@ -267,11 +272,6 @@ namespace Js
                 //anything else should throw a type error
             }
         }
-        else if (args.Info.Count == 1)
-        {
-            //7.1.1 ToPrimitive Note: Date objects treat no hint as if the hint were String.
-            return JavascriptConversion::OrdinaryToPrimitive(args[0], JavascriptHint::HintString/*tryFirst*/, scriptContext);
-        }
 
         JavascriptError::ThrowTypeError(scriptContext, JSERR_FunctionArgument_Invalid, _u("Date[Symbol.toPrimitive]"));
```
Comment (akroshg) [inline lib/Runtime/Library/JavascriptDate.cpp]: nit: why not better error : no hint supplied. or expected hint. something like that.

File: test/es6/toPrimitive.js [hunk 3]
```diff
@@ -159,6 +215,283 @@ var tests = [
...
+            var o = {}, o1 = {}, o2 = {};
+
+            var retVal, hint;
+
+            o[Symbol.toPrimitive] = function(h) { hint.push(h); return retVal; }
+            o1[Symbol.toPrimitive] = function(h) { hint.push("o1:"+h); return retVal; }
+            o2[Symbol.toPrimitive] = function(h) { hint.push("o2:"+h); return retVal; }
+
+            // Ensuring OrdinaryToPrimitive is not called
+            var desc_toString = Object.getOwnPropertyDescriptor(o, "toString");
```
Comment (akroshg) [inline test/es6/toPrimitive.js]: did you use the desc_toString var anywhere in this file.

File: test/es6/toPrimitive.js [hunk 4]
```diff
@@ -159,6 +215,283 @@ var tests = [
...
+
+                //
+                // Date.prototype[@@toPrimitive]
+                //
+
+                assert.throws(()=>(Date.prototype[Symbol.toPrimitive].call(o)), TypeError, "Date.prototype[@@toPrimitive] hint==undefined", "Date[Symbol.toPrimitive]: invalid argument");
+                assert.throws(()=>(Date.prototype[Symbol.toPrimitive].call(o, {})), TypeError, "Date.prototype[@@toPrimitive] hint=={}", "Date[Symbol.toPrimitive]: invalid argument");
+            });
+        }
+    },
```
Comment (akroshg) [inline test/es6/toPrimitive.js]: There is EntrySymbolToPrimitive for SIMD type. Good to add test case for that.

**Comment by suwc** [comment]
>             Assert(!CrossSite::NeedMarshalVar(result, requestContext));

Make sense. Moved to latest rev.

---
In reply to: [234661240](https://github.com/Microsoft/ChakraCore/pull/1319#issuecomment-234661240) [](ancestors = 234661240)

---
Refers to: lib/Runtime/Language/JavascriptConversion.cpp:528 in bf80652. [](commit_id = bf8065293e0ef016de1c59275bac3ed82135ce34, deletion_comment = False)

File: test/es6/toPrimitive.js (hunk 1 above)
Comment (suwc) [inline test/es6/toPrimitive.js]: True - removed in latest rev.

---
In reply to: [71949683](https://github.com/Microsoft/ChakraCore/pull/1319#discussion_r71949683) [](ancestors = 71949683)

File: lib/Runtime/Library/JavascriptDate.cpp (hunk 2 above)
Comment (suwc) [inline lib/Runtime/Library/JavascriptDate.cpp]: Sure - updated.

---
In reply to: [71950085](https://github.com/Microsoft/ChakraCore/pull/1319#discussion_r71950085) [](ancestors = 71950085)

File: test/es6/toPrimitive.js (hunk 3 above)
Comment (suwc) [inline test/es6/toPrimitive.js]: No - removed in latest rev.

---
In reply to: [71950652](https://github.com/Microsoft/ChakraCore/pull/1319#discussion_r71950652) [](ancestors = 71950652)

File: test/es6/toPrimitive.js (hunk 4 above)
Comment (suwc) [inline test/es6/toPrimitive.js]: Will make a separate PR for SIMD  type if necessary.

---
In reply to: [71951385](https://github.com/Microsoft/ChakraCore/pull/1319#discussion_r71951385) [](ancestors = 71951385)

**Comment by akroshg** [comment]
:shipit:
